### PR TITLE
interp: increase speed of findSegment

### PR DESCRIPTION
**Pull request by twpayne**
Using `slices.BinarySearch` instead of `sort.Search` increases the speed of `findSegment` by a factor of two and overall performance by about 30%.

```
goos: linux
goarch: amd64
pkg: gonum.org/v1/gonum/interp
cpu: AMD Ryzen 7 5800 8-Core Processor
                            │  old.bench   │              new.bench              │
                            │    sec/op    │   sec/op     vs base                │
FindSegment-16                104.60n ± 1%   50.78n ± 1%  -51.45% (p=0.000 n=10)
NewPiecewiseLinear-16          114.5n ± 5%   112.2n ± 2%        ~ (p=0.109 n=10)
PiecewiseLinearPredict-16     116.00n ± 1%   84.44n ± 2%  -27.21% (p=0.000 n=10)
PiecewiseConstantPredict-16    87.95n ± 2%   63.93n ± 1%  -27.31% (p=0.000 n=10)
geomean                        105.2n        74.47n       -29.18%
```

Please take a look.

Edit: Markdown formatting.

### CODE REVIEW

File: interp/interp.go [hunk 1]
```diff
@@ -159,7 +159,11 @@ func (pc PiecewiseConstant) Predict(x float64) float64 {
 // is assumed to be +Inf. If no such i is found, it returns -1. It assumes that len(xs) >= 2
```
Comment (kortschak) [inline interp/interp.go:159]: The comment here is incorrect (has always been). I think we need to figure out whether the code is correct or the comment is.

Comment (twpayne) [reply interp/interp.go]: I've removed the erroneous comment and added tests for the edge cases to verify that the code is working as described in the comment.

File: interp/interp_test.go [hunk 2]
```diff
@@ -50,6 +50,22 @@ func TestFindSegment(t *testing.T) {
...
+	if got, want := findSegment([]float64{0}, -1), -1; got != want {
+		t.Errorf("unexpected value of findSegment([]float64{0}, -1): got %d want: %d", got, want)
+	}
+	if got, want := findSegment([]float64{0}, 0), 0; got != want {
+		t.Errorf("unexpected value of findSegment([]float64{0}, 0): got %d want: %d", got, want)
+	}
+	if got, want := findSegment([]float64{0}, 1), 0; got != want {
+		t.Errorf("unexpected value of findSegment([]float64{0}, 1): got %d want: %d", got, want)
+	}
+}
```
Comment (kortschak) [inline interp/interp_test.go:67]:
```suggestion
func TestFindSegmentEdgeCases(t *testing.T) {
	t.Parallel()

	cases := []struct {
		xs   []float64
		x    float64
		want int
	}{
		{xs: nil, x: 0, want: -1},
		{xs: []float64{0}, x: -1, want: -1},
		{xs: []float64{0}, x: 0, want: 0},
		{xs: []float64{0}, x: 1, want: 0},
	}

	for _, test := range cases {
		if got := findSegment(test.xs, test.x); got != test.want {
			t.Errorf("unexpected value of findSegment(%v, %f): got %d want: %d",
				test.xs, test.x, got, test.want)
		}
	}
}
```

**Review by kortschak — COMMENTED** [review]
Prefer this.

**Review by kortschak — APPROVED** [review]
Thanks

**Comment by ayuy5910** [comment]
ありがとう😉👍️🎶

**Comment by ayuy5910** [comment]
難しい
